### PR TITLE
Refactor \MailPoet\Models\Newsletter::getArchives() to Doctrine [MAILPOET-3629]

### PR DESCRIPTION
**Pull request by rodrigoprimo**
One thing to note is that by changing getArchives() to use Doctrine instead of Paris is that I'm changing the type of the variables that are passed to the filters mailpoet_archive_date and mailpoet_archive_subject. I'm unsure what we normally do in those cases. Should I create some code to make sure that this change is backward compatible? Should I change the name of the filter so that at least we don't cause a fatal error for sites that are using one of those filters and deprecate the old filters?

[MAILPOET-3629]

[MAILPOET-3629]: https://mailpoet.atlassian.net/browse/MAILPOET-3629

### CODE REVIEW

**Review by costasovo — COMMENTED** [review]
Good job!

> Should I create some code to make sure that this change is backward compatible? Should I change the name of the filter so that at least we don't cause a fatal error for sites that are using one of those filters and deprecate the old filters?

Very good questions! 

We had a similar issue with the custom shortcode in the newsletter.
What we did back then was that we added some [magic](https://github.com/mailpoet/mailpoet/blob/e59992cc1635a44b96564203a520fc32c9190e34/lib/Entities/NewsletterEntity.php#L163-L173) [getters](https://github.com/mailpoet/mailpoet/blob/a6b58e501accccbf561c8274d7eb52f98e07f65e/lib/Entities/SubscriberEntity.php#L157-L167) on the entities that we started using instead of old models. That way the entity is somewhat backward compatible with the old model.

We also added [an admin notice](https://github.com/mailpoet/mailpoet/blob/65b834a9fff72b1ec5fc81ac383ebb27321da9dc/lib/Util/Notices/DeprecatedShortcodeNotice.php#L1) that we displayed to all users who use the filters. The plan was that after some time we remove those getters and also the notice.

One option is that we could do the same also in this case. We already have magic getters so I guess we could just update comments and prolong for how long we need to keep them. And it would require adding additional notice for these shortcodes.

The backward compatibility is not 100% and we can still cause issues in case someone used type hints for the old models or used some other static methods from the old ORM.

We used this approach in a situation when we already released those BC incompatible changes. We can think of other approaches.

I like the idea of renaming those shortcodes in combination with a notice for users who use the old filter. That way we could eliminate the risk of causing fatal errors. But on the other hand, we would turn off customizations that users made using those filters.

What approach would you prefer?

**Comment by costasovo** [comment]
For additional context, here is a Jira ticket for the previous BC fix: https://mailpoet.atlassian.net/browse/MAILPOET-3333

File: lib/Config/Shortcodes.php [hunk 1]
```diff
@@ -127,10 +135,17 @@ public function getArchive($params) {
       }
       $html .= '<ul class="mailpoet_archive">';
       foreach ($newsletters as $newsletter) {
-        $queue = $newsletter->queue()->findOne();
+        $queue = $newsletter->getQueues()->first();
```
[the remark this answers is not in the document]
Comment (rodrigoprimo) [reply lib/Config/Shortcodes.php]: Thanks, I replaced $newsletter->getQueues()->first() with $newsletter->getLatestQueue().

File: lib/Config/Shortcodes.php [hunk 2]
```diff
@@ -127,10 +135,17 @@ public function getArchive($params) {
...
+        if ($queue instanceof SendingQueueEntity) {
+          $task = $queue->getTask();
+        } else {
+          $task = null;
+        }
+
         $html .= '<li>' .
           '<span class="mailpoet_archive_date">' .
-            $this->wp->applyFilters('mailpoet_archive_date', $newsletter) .
+            $this->wp->applyFilters('mailpoet_archive_date', $task) .
```
[the remark this answers is not in the document]
Comment (rodrigoprimo) [reply lib/Config/Shortcodes.php]: Even though I changed the filter name, I opted to implement your suggestion here and I created the getProcessedAt() proxy method. I believe it is more useful for those using this filter to have an instance of NewsletterEntity than an instance of ScheduledTaskEntity.

**Comment by rodrigoprimo** [comment]
Thanks for your review and your suggestions, @costasovo. I pushed a few new commits implementing the changes that you mentioned. I also pushed a commit creating two new filters and deprecating the old ones.

The downside of this approach, as you mentioned, is that those using this filter will lose any customizations that they might have made. However, the benefit is that we make sure that there is no way someone using the old filter gets a fatal error if they are doing something that is not covered by the backward compatibility code that was added to the NewsletterEntity and SendingQueueEntity classes.

This approach also seemed to me to be the easiest one to implement. Since I believe the impact on those using the old filters is minimal, I don't think it is worth spending too much time creating a more complex solution.

What do you think?

**Review by costasovo — APPROVED** [review]
Thank you! I think this is a very good solution!

I was not aware of the [_deprecated_hook](https://developer.wordpress.org/reference/functions/_deprecated_hook/) function. TIL. There is a notice on the documentation page saying: "This function’s access is marked private. This means it is not intended for use by plugin or theme developers, only in other core functions. It is listed here for completeness." but since we will remove the call soon I think it is not an issue. Am I right?

**Comment by rodrigoprimo** [comment]
> but since we will remove the call soon I think it is not an issue. Am I right?

Yeah, that was my reasoning for using it. Thanks for checking. I didn't use apply_filters_deprecated() which is generally the best approach, because in this particular case I didn't want to keep the old code.

**Review by veljkho — APPROVED** [review]

**Pull request merged**